port/src/devicetree.rs

```diff
@@ -0,0 +1,738 @@
+use core::ffi::CStr;
+
+#[derive(Debug)]
+pub enum ParseError {
+    InvalidHeader,
+    InvalidMagic,
+    BufferTooSmall,
+    InvalidToken,
+}
+
+type Result<T> = core::result::Result<T, ParseError>;
+
+fn bytes_to_u32(bytes: &[u8]) -> Option<u32> {
+    Some(u32::from_be_bytes(bytes.get(..4)?.try_into().unwrap()))
+}
+
+fn bytes_to_u64(bytes: &[u8]) -> Option<u64> {
+    Some(u64::from_be_bytes(bytes.get(..8)?.try_into().unwrap()))
+}
+
+/// Extract u32 from bytes, but cast as u64
+fn bytes_to_u32_as_u64(bytes: &[u8]) -> Option<u64> {
+    Some(u32::from_be_bytes(bytes.get(..4)?.try_into().unwrap()) as u64)
+}
+
+fn align4(n: usize) -> usize {
+    let mut aligned_n = n;
+    let rem = n % 4;
+    if rem > 0 {
+        aligned_n += 4 - rem;
+    }
+    aligned_n
+}
+
+/// DeviceTree is the class entrypoint to the Devicetree operations.
+/// This code focuses only on parsing a Flattened Devicetree without using the heap.
+/// The Devicetree specification can be found here:
+/// https://www.devicetree.org/specifications/
+#[derive(Debug)]
+pub struct DeviceTree<'a> {
+    data: &'a [u8],    // Reference to the underlying data in memory
+    header: FdtHeader, // Parsed structure of the header
+}
+
+impl<'a> DeviceTree<'a> {
+    /// Create new DeviceTree based on memory pointed to by data.
+    /// Result is error if the header can't be parsed correctly.
+    pub fn new(data: &'a [u8]) -> Result<Self> {
+        FdtHeader::new(data, false).map(|header| Self { data, header })
+    }
+
+    /// Given a pointer to the dtb as a u64, return a DeviceTree struct.
+    pub fn from_u64(ptr: u64) -> Result<Self> {
```

Comment:
This function needs to be marked `unsafe`, as it takes an arbitrary integer dereferences a (potentially invalid!) memory address based on it.